src/ds/champ_map_serializers.h

```diff
@@ -0,0 +1,139 @@
+    Version version;
+    V value;
+
+    VersionV() = default;
+    VersionV(Version ver, V val) : version(ver), value(val) {}
+  };
+
+  namespace serialisers
+  {
+    using SerialisedEntry = llvm_vecsmall::SmallVector<uint8_t, 8>;
+  }
+
+  namespace untyped
+  {
+    using SerialisedEntry = champ::serialisers::SerialisedEntry;
+    using VersionV = champ::VersionV<SerialisedEntry>;
+  }
+
+  template <class T>
+  inline size_t get_size(const T& data)
+  {
+    return sizeof(uint64_t) + sizeof(T);
+  }
+
+  template <>
+  inline size_t get_size<champ::untyped::SerialisedEntry>(
+    const champ::untyped::SerialisedEntry& data)
+  {
+    return sizeof(uint64_t) + data.size();
+  }
+
+  template <>
+  inline size_t get_size<champ::untyped::VersionV>(
+    const champ::untyped::VersionV& data)
+  {
+    return sizeof(uint64_t) + sizeof(data.version) + data.value.size();
+  }
+
+  template <class T>
+  inline size_t serialize(const T& t, uint8_t*& data, size_t& size)
+  {
+    uint64_t data_size = sizeof(T);
+    serialized::write(
+      data, size, reinterpret_cast<const uint8_t*>(&data_size), sizeof(T));
+    serialized::write(
+      data, size, reinterpret_cast<const uint8_t*>(&t), sizeof(T));
+    return sizeof(uint64_t) + sizeof(T);
+  }
+
+  template <>
+  inline size_t serialize<champ::untyped::SerialisedEntry>(
+    const champ::untyped::SerialisedEntry& t, uint8_t*& data, size_t& size)
+  {
+    uint64_t data_size = t.size();
+    serialized::write(
+      data,
+      size,
+      reinterpret_cast<const uint8_t*>(&data_size),
+      sizeof(uint64_t));
+    serialized::write(
+      data, size, reinterpret_cast<const uint8_t*>(t.data()), data_size);
+    return sizeof(uint64_t) + data_size;
+  }
+
+  template <>
+  inline size_t serialize<champ::untyped::VersionV>(
+    const champ::untyped::VersionV& t, uint8_t*& data, size_t& size)
+  {
+    uint64_t data_size = sizeof(t.version) + t.value.size();
+    serialized::write(
+      data,
+      size,
+      reinterpret_cast<const uint8_t*>(&data_size),
+      sizeof(uint64_t));
+    serialized::write(
+      data,
+      size,
+      reinterpret_cast<const uint8_t*>(&t.version),
+      sizeof(t.version));
+    serialized::write(
+      data,
+      size,
+      reinterpret_cast<const uint8_t*>(t.value.data()),
+      t.value.size());
+    return sizeof(uint64_t) + sizeof(t.version) + t.value.size();
+  }
+
+  template <class T>
+  inline T deserialize(const uint8_t*& data, size_t& size)
+  {
+    size_t result = serialized::read<size_t>(data, size);
+    CCF_ASSERT_FMT(
+      result == sizeof(T), "result:{} == sizeof(T):{}", result, sizeof(T));
+    return serialized::read<T>(data, size);
+  }
+
+  template <>
+  inline champ::untyped::SerialisedEntry deserialize<
+    champ::untyped::SerialisedEntry>(const uint8_t*& data, size_t& size)
+  {
+    uint64_t data_size = serialized::read<uint64_t>(data, size);
+    champ::untyped::SerialisedEntry ret;
+    ret.assign(data_size, *data);
```

Comment:
Mentioned in call, think this should be something like
```suggestion
    champ::untyped::SerialisedEntry ret(data_size);
    ret.assign(data, data + data_size);
```